### DISCUSSION TITLE
Multi Press and Long Press with Sonoff RF Bridge

### DISCUSSION THREAD

**Discussion by BlueWings172** [arendst/Tasmota#15649] [Q&A] [answered]
Hello

I have a Sonoff RF Bridge that I successfully flashed with Tasmota and Portisch. I have 2 remotes and I've created rules to publish MQTT message when a specific button is pressed.

I was wondering If there is a way to create a rule that works with the RF Bridge and that is triggered by multi press or long press on the remote buttons. In other words, I want to assign a different function to the remote button depending how many times I press it within a timeframe.

Thanks

**Comment by ascillato** [comment, marked as answer]
Example rule to trigger different commands by "1 press of an RF button" or by "keeping pressing for more than 1 second the same button":

```JS
Rule1
    ON Rfreceived#Data=777782 DO Backlog Publish2 home/alarm/set ARM_AWAY; Add1 1; Ruletimer1 3 ENDON
    ON Rules#Timer=1 DO Var1 0 ENDON
    ON Var1#State>1 DO Backlog Var1 0; Publish home/alarm/set TRIGGER ENDON
```

If you press it once, it will publish by MQTT the payload ARM_AWAY
If you press and hold, it will publish by MQTT the payload ARM_AWAY and then TRIGGER

The idea of this example rule is that the button, that gives the RF code 777782, is used as ARM_AWAY for an alarm system, but if you keep it pressed, it will act as a panic button triggering the alarm.

**Reply by ascillato** [reply]
Multi-press is very difficult to properly implement it due to the fact that many **_RF remote keys_** send the same RF code no matter if you press the button once or if you press it twice in the same second. In other words, if you keep a button pressed, you see the same code over and over every half a second or so in the RF Bridge console. Besides, if you press it twice in the same second, you will see the exact same behavior in the RF bridge. So, there is no easy way of knowing if it was a press & hold or a multi-press using common RF remote keys.

**Reply by sfromis** [reply]
Additionally, I think that duplicate signals within a short time will be filtered by the RF reception. And due to the risk of interference, having multipress logic depend on the outdated and not very good RF 433 technology is not really to be recommended anyway.

**Reply by BlueWings172** [reply]
@ascillato 
This worked like magic !! Thanks so much for your help. I don't entirely understand the code but I was able to adapt it for my application and now I can get 2 functions from each button of the remote. 

PS: sorry for the late reply, I couldn't test any of this because I had to rebuild my raspberry pi and container stack after Octoprint failed to update.

**Reply by BlueWings172** [reply]
@ascillato 

There is an issue that I'm facing with the long press:

I have an 8-button remote. I programmed 2 buttons to each haver a completely different function and a long press in each button to cancel that function.

However, when I do a long press on any of the two button, both functions are cancelled ! I have renamed var1 to var5 in one of the rules but still when I long-press any button it triggers two actions in two different rules !?!

would appreciate any help on this.

thanks

**Reply by sfromis** [reply]
If you rename Var1, you should do it in all 3 rules for the same button, including the Add1 command. And you should also have a separate RuleTimer for each. If this was not it, better post what you *exactly* did....

**Reply by ascillato** [reply]
Please, post all your rules.

**Reply by BlueWings172** [reply]
@ascillato,  @sfromis 

I understand now that for each rule, I have to have different var, add and realtimer.
So for the first rule I have var1, add1, realtimer1 and for the second rule I renamed them to var2, add2, and realtimer2 and so on.

there is 1 minor issue though.

Long press does't really work. I press once and I trigger action 1 and If I short press again, I trigger action 2. What I mean is that a second normal press is sufficient to trigger action 2 and no long press is required. Long press still works if I use var1, add1, and realtimer1.

This is the code I'm using:

```
Rule3
    ON Rfreceived#Data=F6D0F1 DO Backlog webquery http://192.168.1.10:8089/admin/api.php?disable=300&auth=key GET; Add2 1; Ruletimer2 3 ENDON
    ON Rules#Timer=1 DO var2 0 ENDON
    ON var2#State>1 DO Backlog var2 0; Rfraw 0xC0; webquery http://192.168.1.10:8089/admin/api.php?enable&auth=key GET ENDON
```



I have added Rfraw 0xC0 in action 2 so I hear a beep when action 2 is triggered.

This is what I see in the console when I press once, followed by another short press after roughly 3 seconds:

```
18:10:42.592 MQT: tele/rfbridge/RESULT = {"Time":"2022-05-29T18:10:42","RfReceived":{"Sync":7380,"Low":230,"High":720,"Data":"F6D0F1","RfKey":"None"}}
18:10:42.626 RUL: RFRECEIVED#DATA=F6D0F1 performs "Backlog webquery http://192.168.1.10:8089/admin/api.php?key; Add2 1; Ruletimer2 3"
18:10:42.738 MQT: stat/rfbridge/RESULT = {"WebQuery":"Done"}
18:10:42.879 MQT: stat/rfbridge/RESULT = {"Add2":"1.000"}
18:10:43.090 MQT: stat/rfbridge/RESULT = {"T1":0,"T2":3,"T3":0,"T4":0,"T5":0,"T6":0,"T7":0,"T8":0}
18:10:46.395 MQT: tele/rfbridge/RESULT = {"Time":"2022-05-29T18:10:46","RfReceived":{"Sync":7360,"Low":230,"High":710,"Data":"F6D0F1","RfKey":"None"}}
18:10:46.429 RUL: RFRECEIVED#DATA=F6D0F1 performs "Backlog webquery http://192.168.1.10:8089/admin/api.php?key GET; Add2 1; Ruletimer2 3"
18:10:46.519 MQT: stat/rfbridge/RESULT = {"WebQuery":"Done"}
18:10:46.658 MQT: stat/rfbridge/RESULT = {"Add2":"2.000"}
18:10:46.723 RUL: VAR2#STATE>1 performs "Backlog var2 0; Rfraw 0xC0; webquery http://192.168.1.10:8089/admin/api.php?key GET"
18:10:46.743 MQT: stat/rfbridge/RESULT = {"T1":0,"T2":2,"T3":0,"T4":0,"T5":0,"T6":0,"T7":0,"T8":0}
18:10:47.075 MQT: stat/rfbridge/RESULT = {"Var2":"0"}
18:10:47.288 MQT: stat/rfbridge/RESULT = {"RfRaw":"OFF"}
18:10:47.622 MQT: stat/rfbridge/RESULT = {"WebQuery":"Done"}

```


thanks for your help and sorry for the late reply. I was testing many times and I thought I had other issues but things seems to be working.

**Reply by sfromis** [reply]
You are not using the right RuleTimer here, it should be number 2, but you use number 1
`ON Rules#Timer=1 DO var2 0 ENDON`
Such details is why you should post *exactly* what you do, not just say what you think you do.

**Comment by BlueWings172** [comment]
@sfromis 

Thanks for the clarification.

I changed the value for Rules#Timer but still did not make a difference. I've done many tests and below are the rule i have used and the console output when I press (action1) and then long press (action2) after just very few seconds.

Please note below that i'm using `var4 `,  `add4 ` , `Ruletimer4 ` and `Rules#Timer=4 `

```
rule2
	ON Rfreceived#Data=F6D0F1 DO Backlog webquery http://192.168.1.10:8089/admin/api.php?disable=300&auth=key GET; Add4 1; Ruletimer4 3 ENDON
	ON Rules#Timer=4 DO var4 0 ENDON
	ON var4#State>1 DO Backlog var4 0; Rfraw 0xC0; webquery http://192.168.1.10:8089/admin/api.php?enable&auth=key GET ENDON"

```

Console output:

```
14:12:42.958 MQT: tele/rfbridge/RESULT = {"Time":"2022-05-30T14:12:42","RfReceived":{"Sync":7390,"Low":250,"High":720,"Data":"F6D0F1","RfKey":"None"}}
14:12:42.013 RUL: RFRECEIVED#DATA=F6D0F1 performs "Backlog webquery http://192.168.1.10:8089/admin/api.php?disable=300&auth=key GET; Add4 1; Ruletimer4 3"
14:12:43.130 MQT: stat/rfbridge/RESULT = {"WebQuery":"Done"}
14:12:43.231 MQT: stat/rfbridge/RESULT = {"Add4":"1.000"}
14:12:43.470 MQT: stat/rfbridge/RESULT = {"T1":0,"T2":0,"T3":0,"T4":2,"T5":0,"T6":0,"T7":0,"T8":0}
14:12:46.632 MQT: tele/rfbridge/RESULT = {"Time":"2022-05-30T14:12:46","RfReceived":{"Sync":7390,"Low":240,"High":720,"Data":"F6D0F1","RfKey":"None"}}
14:12:46.689 RUL: RFRECEIVED#DATA=F6D0F1 performs "Backlog webquery http://192.168.1.10:8089/admin/api.php?disable=300&auth=key GET; Add4 1; Ruletimer4 3"
14:12:46.763 MQT: stat/rfbridge/RESULT = {"WebQuery":"Done"}
14:12:46.916 MQT: stat/rfbridge/RESULT = {"Add4":"2.000"}
14:12:46.013 RUL: VAR4#STATE>1 performs "Backlog var4 0; Rfraw 0xC0; webquery http://192.168.1.10:8089/admin/api.php?enable&auth=key GET"
14:12:47.009 MQT: stat/rfbridge/RESULT = {"T1":0,"T2":0,"T3":0,"T4":2,"T5":0,"T6":0,"T7":0,"T8":0}
14:12:47.211 MQT: stat/rfbridge/RESULT = {"Var4":"0"}
14:12:47.450 MQT: stat/rfbridge/RESULT = {"RfRaw":"OFF"}
14:12:47.773 MQT: stat/rfbridge/RESULT = {"WebQuery":"Done"}
14:12:50.167 RUL: RULES#TIMER=4 performs "var4 0"
14:12:50.176 MQT: stat/rfbridge/RESULT = {"Var4":"0"}
```

**Reply by sfromis** [reply]
I'm assuming that your issue is about what webquery URL's you get pinged.... What I see is two with disable=300 for the two button pushes, which looks very consistent with the rules. However, the 3rd URL looks "surprising" to me, as I'm assuming that the id 300 is different per channel? `http://192.168.1.10:8089/admin/api.php?enable&auth=key`
That URL has no channel (or what you call it) number, should it have been `enable=300` ?

**Reply by sfromis** [reply]
Of potential interest about how fast you can get results of pushing buttons.... https://github.com/arendst/Tasmota/issues/10248

**Reply by BlueWings172** [reply]
@sfromis 

I've done some tests and finally restarted the rfbridge. I think it works now. What is discovered is that if I short press again within 3 seconds or so, then action 2 is triggered without the need of a long press (i don't think this is a frequent case in the real world). However, if I short press after 10 seconds or so then action1 is triggered (like it should) and I need to long press to trigger action2. So the issue could appear only within the first few seconds.

i'm glad I added the beep to action 2 and I advice anyone who is going to implement long press actions to include a beep for action2 in order to be able to tell which action is being triggered.

I wish I was able to understand how the code works. This would've helped me determine what the issue was but I'm very new to this and i will have to spend some time with the documentation and hopefully decipher the code and understand the logic behind it.

As to the WebQuery, I'm certain that it works because i've tested it in the browser and console. It basically disables pihole for 300 seconds (this is action1) and if I long press, it should enable pihole again. I have not included the api key because it is too long and makes reading the console output very difficult.

I'm super happy now with my remote now. The RFbridge is underestimated in my opinion and can do lots of great things.

Thanks so much mate for the help and support.
